ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/HD/BackingStore.hs

```diff
@@ -0,0 +1,987 @@
+    fmap DS.NEDiffHistory $ mapMaybe NESeq.nonEmptySeq $ QC.shrink (NESeq.toSeq h)
+
+deriving newtype instance (Ord k, QC.Arbitrary k, QC.Arbitrary v)
+                       => QC.Arbitrary (DS.Values k v)
+
+deriving newtype instance (Ord k, QC.Arbitrary k)
+                       => QC.Arbitrary (DS.Keys k v)
+
+deriving newtype instance (Ord k, QC.Arbitrary k, QC.Arbitrary v)
+                       => QC.Arbitrary (DS.Diff k v)
+
+instance (Ord k, QC.Arbitrary k)
+      => QC.Arbitrary (KeysMK k v) where
+  arbitrary = ApplyKeysMK <$> QC.arbitrary
+  shrink (ApplyKeysMK ks) = ApplyKeysMK <$> QC.shrink ks
+
+instance (Ord k, QC.Arbitrary k, QC.Arbitrary v)
+      => QC.Arbitrary (DiffMK k v) where
+  arbitrary = ApplyDiffMK <$> QC.arbitrary
+  shrink (ApplyDiffMK d)= ApplyDiffMK <$> QC.shrink d
+
+deriving newtype instance QC.Arbitrary (ApplyMapKind' mk k v)
+                       => QC.Arbitrary (
+                            LedgerTables
+                              (SimpleLedgerState k v)
+                              (ApplyMapKind' mk)
+                            )
+
+{------------------------------------------------------------------------------
+  Labelling
+------------------------------------------------------------------------------}
+
+data Tag =
+    SimpleWrite
+  deriving (Show, Eq)
+
+type EventPred keys values diff =
+  Label.Predicate (Event keys values diff Symbolic) Tag
+
+tagEvents ::
+     forall keys values diff.
+     Operations keys values diff
+  -> [Event keys values diff Symbolic]
+  -> [Tag]
+tagEvents _ops = Label.classify [
+      tagSimpleWrite
+    ]
+  where
+    tagSimpleWrite :: EventPred keys values diff
+    tagSimpleWrite = Label.predicate go
+      where
+        go ev = case (cmd ev, mockResp ev) of
+          (At (BSWrite _ _), Resp (Right (Unit ()))) ->
+            Left SimpleWrite
+          _ ->
+            Right tagSimpleWrite
+
+execCmd ::
+     Operations keys values diff
+  -> Model values Symbolic
+  -> QSM.Command (At (Cmd keys values diff)) (At (Resp keys values diff))
+  -> Event keys values diff Symbolic
+execCmd ops model (QSM.Command cmd resp _vars) =
+    lockstep ops model cmd resp
+
+execCmds ::
+     forall keys values diff.
+     Operations keys values diff
+  -> QSM.Commands (At (Cmd keys values diff)) (At (Resp keys values diff))
+  -> [Event keys values diff Symbolic]
+execCmds ops = \(QSM.Commands cs) -> go (initModel ops) cs
+  where
+    go ::
+         Model values Symbolic
+      -> [QSM.Command (At (Cmd keys values diff)) (At (Resp keys values diff))]
+      -> [Event keys values diff Symbolic]
+    go _ []       = []
+    go m (c : cs) = e : go (after e) cs
+      where
+        e = execCmd ops m c
+
+{------------------------------------------------------------------------------
+  Inspecting the labelling function
+------------------------------------------------------------------------------}
+
+showLabelledExamples :: Maybe Int -> IO ()
```

Comment:
Are we using this? If not we should do some coverage testing to make sure we test the important cases.